src/Ouroboros/Storage/Immutable/FS.hs

```diff
@@ -0,0 +1,900 @@
+children (FolderOnDisk _ c) = c
+
+-- "/var/foo.txt"
+-- index ["/", "var"] ..
+--   FolderOnDisk "var" [FileOnDisk "foo" []]
+-- index ["foo.txt"] F
+
+data MockFS = MockFS {
+    getMockFS :: FsTree ByteString String
+  } deriving Show
+
+-- | Renders the 'MockFS' in a human-readable fashion.
+prettyShowFS :: MockFS -> String
+prettyShowFS MockFS{..} =
+    let prettyDisk =
+            map (\(fp, disk) -> "- " <> fp
+                                     <> " -> "
+                                     <> show (hexDump $ B16.encode disk)
+                                     <> " (" <> show disk <> ")")
+                (collectFiles getMockFS)
+    in L.intercalate "\n" prettyDisk <> "\n\n" <> drawTree getMockFS
+    where
+        collectFiles :: FsTree ByteString String -> [(FileName, ByteString)]
+        collectFiles (FolderOnDisk _ ts) =
+            mconcat (map collectFiles ts)
+        collectFiles (FileOnDisk f b) = [(f,b)]
+
+        hexDump :: ByteString -> ByteString
+        hexDump = fst
+                . BS.foldl' (\(acc, n) w8 ->
+                                if n == 2 then (acc <> " " <> BS.singleton w8, 1)
+                                          else (acc <> BS.singleton w8, n + 1)
+                            ) (mempty, 0 :: Int)
+
+
+drawTree :: FsTree a String -> String
+drawTree  = unlines . draw
+  where
+    draw :: FsTree a String -> [String]
+    draw (FileOnDisk f _)     = ["(f) " <> f]
+    draw (FolderOnDisk x ts0) = lines x ++ drawSubTrees ts0
+      where
+        drawSubTrees [] = []
+        drawSubTrees [t] =
+            "|" : shiftT "`- " "   " (draw t)
+        drawSubTrees (t:ts) =
+            "|" : shiftT "+- " "|  " (draw t) ++ drawSubTrees ts
+
+        shiftT first other = zipWith (++) (first : repeat other)
+
+-- | Given a 'FilePath' returns a new 'FsTree' populated with all the
+-- directories listed in the input 'FilePath'.
+fsFromFilePath :: HasCallStack => FilePath -> FsTree a String
+fsFromFilePath fp =
+    case splitDirectories fp of
+        [] -> error "fsFromFilePath: empty input path."
+        ps -> mockCreateDirectoryIfMissingInternal True ps (FolderOnDisk (head ps) [])
+
+demoMockFS :: MockFS
+demoMockFS = MockFS {
+    getMockFS = FolderOnDisk "/" [
+      FolderOnDisk "usr" [
+                   FolderOnDisk "local" [FolderOnDisk "bin" mempty]
+                 ]
+    , FolderOnDisk "var" [
+                   FolderOnDisk "log"  mempty
+                 , FolderOnDisk "tmp"  [FileOnDisk "foo.txt" mempty]
+                 , FolderOnDisk "mail" mempty
+                 , FolderOnDisk "run"  mempty
+                 ]
+    ]
+  }
+
+newEmptyMockFS :: MockFS
+newEmptyMockFS = MockFS {
+    getMockFS = FolderOnDisk "/" []
+  }
+
+-- Like 'newEmptyMockFS', but allows the user to supply an input 'FilePath' to
+-- create an initial directory on disk.
+newMockFsWith :: FilePath -> MockFS
+newMockFsWith dir = newEmptyMockFS { getMockFS = fsFromFilePath dir }
+
+data SimFsAction m =
```

Comment:
This needs a comment explaining what this is for.

Reply:
In fact, let's comment it out, remove all uses of it, but add a comment saying it might be a useful starting point for generating test scripts.